contributors/design-proposals/namespace-specific-dns.md

```diff
@@ -0,0 +1,55 @@
+
+Once `--dns-namespace` flag is provided to `kube-dns`, instead of monitoring the changes from the whole cluster’s kubernetes services/endpoints, this single DNS server will only monitor the changes of kubernetes services/endpoints that belong to the in-scope namespaces —— dns-namespace & default-namespace. Therefore, we can modify the implementation of `#setServicesStore` and `#setEndpointsStore` to register the callback methods only for the services/endpoints from specific namespaces. As a result, instead of just creating one pair of `cache.Store` and `cache.Controller` objects to list & watch resource changes from all the namespaces in `#setServicesStore` and `#setEndpointsStore` respectively, we would create two pairs of the store and controller objects under this situation. 1 pair of the store and controller objects lists & watches the resource changes from `default` namespace, while the other pair monitors the resource changes from the namespace defined by `--dns-namespace`.
+
+Accordingly, the `#Start` method now will need to start 4 go routines for the controllers. 2 controllers to watch service changes from default-namespace and dns-namespace, while 2 other controllers for endpoint changes.
+
+As a result, if the above changes are made to KubeDNS and launch multi-DNS server with the `--dns-namespace` flag, each single DNS server will only provide DNS service for the kubernetes services from restricted/scoped namespaces.
+
+There is already some POC code of this proposal available [here](https://github.com/duglin/kubernetes/commit/15fbe749cb8df7a637e1365059ac987b9c89a089).
+
+* `--dns-namespace` flag is not provided to KubeDNS
+
+Under this situation, the behavior of KubeDNS should remain the same as before, which means, `#setServicesStore` and `#setEndpointsStore` should only create the controllers to monitor the resource changes from all the namespaces.
+
+* Namespace specific DNS client config
+
+Once we enable a namespace specific DNS server in a kubernetes cluster, we shall make kubelet aware of multiple DNS servers. When creating pods from different namespaces, kubelet can add different DNS client config to let different pods point to different DNS servers.
+
+To implement this, kubernetes should allow namespace rather than kubelet to supply the added DNS client config. We already submitted a proposal for adding this feature in K8S community, please refer to this [link](https://github.com/kubernetes/features/issues/75) to check the details of the design, and also the patch of the changes is available [here](https://github.com/ardnaxelarak/kubernetes/commit/53ec8dbae38b15e4236996dc7857b19648c64b7b).
```

Comment:
I want to support the use case where a k8s operator offers a multi-tenant k8s service in which a tenant can run standard examples completely, or as near as possible, un-modified.  The standard examples do not say `NamespaceFirst`.  Requiring only an annotation on the tenant's Namespace is the smallest change I can think of.